### PR TITLE
Self documenting f-strings

### LINKED ISSUE #3796: [BUG] self documenting expressions in f-strings
**Describe the bug**

Python 3.8 added support for the `=` specifier in f-strings to provide self documenting expressions. Using this feature makes a file fail to compile.

**To Reproduce**

test.py:
```test.py
x = 1
print(f"{x=}")
```

then run:

```
$ python test.py
x=1

$ cython -3 test.py

Error compiling Cython file:
------------------------------------------------------------
...
x = 1
print(f"{x=}")
            ^
------------------------------------------------------------

test.py:2:13: Expected ')', found '='
```

**Expected behavior**
No error compiling the file

**Environment (please complete the following information):**
 - OS: macOS
 - Python version 3.8.3
 - Cython version 0.29.21

**Additional context**

https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging

### PR DESCRIPTION
**Pull request by davfsa**
Fixes #3796

### CODE REVIEW

**Comment by da-woods** [comment]
There's some examples in "test/run/test_fstring.pyx" that are currently commented out

https://github.com/cython/cython/blob/cedbd5840fde2c63549b64c35cd05ea5fb9fc524/tests/run/test_fstring.pyx#L1125-L1229

They're stolen from the CPython test-suite I believe. Uncommenting them should be fine for this PR.

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -3340,14 +3340,16 @@ class FormattedValueNode(ExprNode):
     # {}-delimited portions of an f-string
     #
     # value           ExprNode                The expression itself
-    # conversion_char str or None             Type conversion (!s, !r, !a, or none, or 'd' for integer conversion)
+    # conversion_char str or None             Type conversion (!s, !r, !a, none, or 'd' for integer conversion)
     # format_spec     JoinedStrNode or None   Format string passed to __format__
+    # expr_text       str or NOne             If not None, the expression text
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:3345]: a) "NOne" -> None
b) Maybe a clearer name (or at least a clearer comment) - `self_documenting_text`?

Comment (davfsa) [reply Cython/Compiler/ExprNodes.py:3345]: Will fix that `None` and fix the comment :+1:

**Comment by davfsa** [comment]
> There's some examples in "test/run/test_fstring.pyx" that are currently commented out
> 
> https://github.com/cython/cython/blob/cedbd5840fde2c63549b64c35cd05ea5fb9fc524/tests/run/test_fstring.pyx#L1125-L1229
> 
> They're stolen from the CPython test-suite I believe. Uncommenting them should be fine for this PR.

Oh, so you store them on `.pyx` to compile them, should have thought about that one (got overwhelmed by so many tests and didnt know where to start :sweat_smile:). Thanks for the heads up

**Comment by davfsa** [comment]
@da-woods Fixed. Btw, you guys dont seem to have auto-cancellation of builds on pr's, which might be worth turning on specially due to the pipelines being limited to 1 concurrently and taking really long to run.
image

**Comment by da-woods** [comment]
`python3 runtests.py test_fstrings` should at least let you check that that specific test runs on your PC.

File: Cython/Compiler/ExprNodes.py [hunk 2]
```diff
@@ -3420,6 +3422,13 @@ def generate_result_code(self, code):
             code.globalstate.use_utility_code(
                 UtilityCode.load_cached("PyObjectFormatSimple", "StringTools.c"))
 
+        if self.expr_text:
+            expr_text = self.expr_text + '='
+            value_result = 'PyNumber_Add(PyUnicode_FromStringAndSize("%s", %s), %s)' % (
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:3427]: I'm pretty sure this is leaking the result of `PyUnicode_FromStringAndSize`, and probably the result of `PyNumber_Add` too.

I think this might have to be a new section in `Cython/Utility/Stringtools.c` similar to `PyObjectFormatAndDecref` etc. But this is the first time I've looked at the fstring implementation so I'm not complete confident about the best way to do it.

Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:3427]: An alternative approach that might be worth a go is to generate a different set of nodes (probably at the `analyse_ expressions` stage). Something like:

```
# in analyse_expressions for FormattedValueNode:
if self.expr_text:
    return AddNode(self.pos, lhs=UnicodeConstantNode(self.pos, value=self.expr_text+"="),
                                       rhs=<copy of self but without self.expr_text>).analyse_expressions(env)
```

The advantage of doing something like that is that Cython's code generation should take care of most of the details. Note that the code above is a very vague outline...

Comment (davfsa) [reply Cython/Compiler/ExprNodes.py:3427]: Will have a try with that approach :+1: 

Btw, mind canceling all the old builds? They are taking for ever to complete:
https://travis-ci.org/github/cython/cython/builds/748578596
https://travis-ci.org/github/cython/cython/builds/748556005

Thanks again!

Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:3427]: > Btw, mind canceling all the old builds?

I don't have the power I'm afraid.

Comment (davfsa) [reply Cython/Compiler/ExprNodes.py:3427]: Oh, thought you did, nvm then :+1:

Comment (davfsa) [reply Cython/Compiler/ExprNodes.py]: Implemented :smile:

**Comment by da-woods** [comment]
It looks reasonable to me (although my knowledge of the parser is a little vague)

**Comment by davfsa** [comment]
> It looks reasonable to me (although my knowledge of the parser is a little vague so wouldn't like to comment too much on that bit)

Ok, thanks for having a look :+1:

**Comment by davfsa** [comment]
Little bump :) @scoder

File: Cython/Compiler/Parsing.py [hunk 3]
```diff
@@ -1161,6 +1168,16 @@ def p_f_string_expr(s, unicode_value, pos, starting_index, is_raw):
         error(_f_string_error_pos(pos, unicode_value, starting_index),
               "empty expression not allowed in f-string")
 
+    if terminal_char == '=':
+        i += 1
+        while unicode_value[i] == ' ':
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:1173]: What does CPython do here? `' '` or `.isspace()`?

File: Cython/Compiler/ExprNodes.py [hunk 4]
```diff
@@ -3377,6 +3380,12 @@ def analyse_types(self, env):
                 if self.value.type is unicode_type and not self.value.may_be_none():
                     # value is definitely a unicode string and we don't format it any special
                     return self.value
+
+        if self.expr_text:
+            # strings are unicode in python, so we can safely do this
+            string_node = StringNode(self.pos, value=StringEncoding.BytesLiteral(self.expr_text.encode('utf-8')))
+            return AddNode(self.pos, operand1=string_node, operand2=self, operator='+', type=string_node.type, is_temp=True)
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:3387]: Why is this not done in the parser? It builds a sequence of strings – I don't see a need for an additional `AddNode`.

**Review by scoder — CHANGES_REQUESTED** [review]

File: Cython/Compiler/Parsing.py (hunk 3 above)
Comment (davfsa) [reply Cython/Compiler/Parsing.py:1173]: They seem to do `.isspace()`. Will change

https://github.com/python/cpython/pull/13123/files#diff-63309d32b22eaebd869b7ac0d2d4d02f95bd7ff5de2ba042d07c8e6962af8bbcR5206

File: Cython/Compiler/ExprNodes.py (hunk 4 above)
Comment (davfsa) [reply Cython/Compiler/ExprNodes.py:3387]: This was suggested by @da-woods [here](https://github.com/cython/cython/pull/3939#discussion_r539574402) I'm kind of new and still trying to understand how all this works, so I don't know the best way to do this :sweat_smile:

Comment (da-woods) [reply Cython/Compiler/ExprNodes.py:3387]: I have been known to give bad advice.... Although I think this was still an improvement on the code that it replaced.

But yes, given that the `expr_text` is known as you parse it it's probably most efficient to handle it in the parser. Maybe if `p_f_string_expr` returned the index and a sequence instead (e.g. either `[expr_text, FormattedValueNode]` or `[FormattedValueNode]`)?

Comment (scoder) [reply Cython/Compiler/ExprNodes.py:3387]: > I have been known to give bad advice

New code means that we're solving a problem that we haven't solved before (at least not in this context). It's perfectly ok to try a couple of approaches until we understand the problem well enough to find a _good_ way to solve it. Every "bad" approach helps in making the "obvious" solution clearer, until we're able to see it (and then slap out foreheads, commit it, and start coding up our next "bad" approach for the next problem).

File: Cython/Compiler/Parsing.py [hunk 5]
```diff
@@ -1161,6 +1168,16 @@ def p_f_string_expr(s, unicode_value, pos, starting_index, is_raw):
         error(_f_string_error_pos(pos, unicode_value, starting_index),
               "empty expression not allowed in f-string")
 
+    if terminal_char == '=':
+        i += 1
+        while unicode_value[i].isspace():
+            if i >= size:
+                break  # error will be reported below
+            i += 1
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:1176]: This looks like `i` might exceed the length of `unicode_value` here if the string ends with a `=` or spaces. This could trigger a compiler crash rather than an error report, either right in the `while` condition or further down when reading the new `terminal_char`.

File: Cython/Compiler/ExprNodes.py (hunk 4 above)
Comment (davfsa) [reply Cython/Compiler/ExprNodes.py:3387]: > I have been known to give bad advice.

I didnt mean to blame it on you :sweat_smile:. I just wanted to point out where I got the code from so that scoder could have a look at the discussion. I should have worded that better, sorry!

> Maybe if p_f_string_expr returned the index and a sequence instead (e.g. either [expr_text, FormattedValueNode] or [FormattedValueNode])?

Will try that!

File: Cython/Compiler/Parsing.py (hunk 5 above)
Comment (davfsa) [reply Cython/Compiler/Parsing.py:1176]: Ending with an `=` or ending with spaces (or a combination of both) works perfectly fine. I did notice however that a malformed string does cause a compiler crash. Will fix

Code:
```
a = "test"
b = f"{a="
```

Comment (davfsa) [reply Cython/Compiler/Parsing.py]: Fixed

File: Cython/Compiler/ExprNodes.py (hunk 4 above)
Comment (davfsa) [reply Cython/Compiler/ExprNodes.py]: Moved to the parser

File: Cython/Compiler/Parsing.py [hunk 6]
```diff
@@ -1082,7 +1082,7 @@ def p_f_string(s, unicode_value, pos, is_raw):
                     values.append(ExprNodes.UnicodeNode(pos, value=builder.getstring()))
                     builder = StringEncoding.UnicodeLiteralBuilder()
                 next_start, expr_node = p_f_string_expr(s, unicode_value, pos, next_start, is_raw)
-                values.append(expr_node)
+                values.extend(expr_node)
```
Comment (da-woods) [inline Cython/Compiler/Parsing.py:1085]:
```suggestion
                next_start, expr_nodes = p_f_string_expr(s, unicode_value, pos, next_start, is_raw)
                values.extend(expr_nodes)
```

Other than this tiny comment I think this makes sense - it's does look simpler to do this in the parser.

Comment (davfsa) [reply Cython/Compiler/Parsing.py]: Yep, this looks like a much nicer way to do it. Thanks for the suggestion :)

PS: Fixed

File: Cython/Compiler/Parsing.py [hunk 7]
```diff
@@ -1161,6 +1171,21 @@ def p_f_string_expr(s, unicode_value, pos, starting_index, is_raw):
...
+                break  # error will be reported below
+
+            c = unicode_value[i]
+            if c.isspace():
+                i += 1
+                continue
+
+            terminal_char = c
+            expr_text = unicode_value[starting_index:i]
+            break
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:1187]: Let's simplify this a little.
```suggestion
        while i < size and unicode_value[i].isspace():
            i += 1
        if i < size:
            expr_text = unicode_value[starting_index:i]
            terminal_char = unicode_value[i]
        # otherwise: error will be reported below
```

File: tests/run/test_fstring.pyx [hunk 8]
```diff
@@ -1169,9 +1168,10 @@ non-important content
         self.assertEqual(f'{0!=1}', 'True')
         self.assertEqual(f'{0<=1}', 'True')
         self.assertEqual(f'{0>=1}', 'False')
-        self.assertEqual(f'{(x:="5")}', '5')
+        # Walrus not implemented yet, skip
+        #self.assertEqual(f'{(x:="5")}', '5')
+        #self.assertEqual(f'{(x:=5)}', '5')
         self.assertEqual(x, '5')
-        self.assertEqual(f'{(x:=5)}', '5')
```
Comment (scoder) [inline tests/run/test_fstring.pyx:1174]: Let's keep the order here. This file might get overwritten and manually merged from time to time with new copies from CPython's test suite. It's confusing if lines move around on our side.

**Comment by davfsa** [comment]
@scoder both fixed

File: tests/run/test_fstring.pyx [hunk 9]
```diff
@@ -1169,9 +1168,11 @@ non-important content
         self.assertEqual(f'{0!=1}', 'True')
         self.assertEqual(f'{0<=1}', 'True')
         self.assertEqual(f'{0>=1}', 'False')
-        self.assertEqual(f'{(x:="5")}', '5')
+        # Walrus not implemented yet, skip
+        #self.assertEqual(f'{(x:="5")}', '5')
         self.assertEqual(x, '5')
```
Comment (scoder) [inline tests/run/test_fstring.pyx:1173]: This looks like it should probably fail now.

**Review by scoder — APPROVED** [review]

File: tests/run/test_fstring.pyx (hunk 9 above)
Comment (davfsa) [reply tests/run/test_fstring.pyx:1173]: It looks as if the tests never ran due to the `__` in the name of the function. Should I remove them?

Comment (davfsa) [reply tests/run/test_fstring.pyx]: Fixed

**Comment by scoder** [comment]
Looks good now. We need a work-around for Py2.7, though:
```
ERROR: test_debug_conversion (test_fstring.TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/run/test_fstring.pyx", line 1131, in test_fstring.TestCase.test_debug_conversion (test_fstring.c:44342)
    self.assertEqual(f'{x=!a}', 'x=' + ascii(x))
NameError: name 'ascii' is not defined
```

**Comment by davfsa** [comment]
> Looks good now. We need a work-around for Py2.7, though:
> 
> ```
> ERROR: test_debug_conversion (test_fstring.TestCase)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "tests/run/test_fstring.pyx", line 1131, in test_fstring.TestCase.test_debug_conversion (test_fstring.c:44342)
>     self.assertEqual(f'{x=!a}', 'x=' + ascii(x))
> NameError: name 'ascii' is not defined
> ```

I havent been able to replicate `ascii` on python2 or find any backport. The docs state that `repr` produces a similar result in python2, but just to stay safe and avoid edge cases, they will only run on python3.

Source: https://docs.python.org/3/library/functions.html#ascii

File: tests/run/test_fstring.pyx [hunk 10]
```diff
@@ -1122,21 +1122,22 @@ non-important content
...
-    """
-    def __test_debug_conversion(self):
+    def test_debug_conversion(self):
         x = 'A string'
         self.assertEqual(f'{x=}', 'x=' + repr(x))
         self.assertEqual(f'{x =}', 'x =' + repr(x))
         self.assertEqual(f'{x=!s}', 'x=' + str(x))
         self.assertEqual(f'{x=!r}', 'x=' + repr(x))
-        self.assertEqual(f'{x=!a}', 'x=' + ascii(x))
+        if not IS_PY2: # ASCII formatting only works on Python 3+
```
Comment (da-woods) [inline tests/run/test_fstring.pyx:1131]: I the comment is probably wrong - i think `!a` in an f-string works on Cython on Py2, but it's the `ascii` builtin that's missing on Py2 (and thus it's hard to test). Skipping the test may still be reasonable though.

Comment (davfsa) [reply tests/run/test_fstring.pyx]: Re-worded :)

**Comment by davfsa** [comment]
> 
> 
> > I havent been able to replicate `ascii` on python2 or find any backport. The docs state that `repr` produces a similar result in python2, but just to stay safe and avoid edge cases, they will only run on python3.
> 
> We actually map `!a` to `repr()` in Cython for Python 2.7. Just define `ascii = repr` if `IS_Py2` at the top of the file.

Ah, ok, will do 👍

**Comment by davfsa** [comment]
@scoder Small bump on this one, everything should be working

File: Cython/Compiler/Parsing.py [hunk 11]
```diff
@@ -1143,11 +1147,17 @@ def p_f_string_expr(s, unicode_value, pos, starting_index, is_raw):
...
+            # allow special cases with '!' and '='
+            if i + 1 < size:
+                chars = c + unicode_value[i + 1]
+                if chars in ('!=', '==', '>=', '<='):
+                    i += 2  # we checked 2, so we can skip 2
+                    continue
+
+                if c in '><':  # allow single '<' and '>'
+                    i += 1
+                    continue
```
Comment (scoder) [inline Cython/Compiler/Parsing.py:1160]: We can avoid the string building here.
```suggestion
                if c in '!=><':
                    if unicode_value[i + 1] == '=':
                        i += 2  # we checked 2, so we can skip 2
                        continue
                    elif c in '><':  # allow single '<' and '>'
                        i += 1
                        continue
```

**Comment by scoder** [comment]
Let's get this feature merged. Thanks for working it up!
